### PR TITLE
Add paramtype2s for 4 horizontal rotations and 64 colors

### LINKED ISSUE #6194: Feature Request:  simple color facedir
This has been brought up once or twice before but I figured I ought to formalize it here:

I'd like to see a mode added (preferably before 0.4.17 goes out, though I'm sure that's unlikely), that adds a "simple" color-facedir, where one can select one the four cardinal N/S/W/E directions, similar to wallmounted mode, but without the up- and down-facing orientations.

This would require only two bits for the rotation, while the upper six bits would then be used to specify a palette index, in the same manner as the other color modes (but of course, 64 colors in this case).  I imagine this would be adequate for the majority of items that need both rotation and color.

### PR DESCRIPTION
**Pull request by Wuzzy2**
Fixes #6194. @VanessaE. @paramat. @C1ffisme. @Sokomine.

This PR adds 2 paramtype2s: `4dir` and `color4dir`.

`4dir` is like `facedir`, but only for 4 horizontal directions: NESW. It is identical in behavior to `facedir` otherwise. The reason why game makers would want to use this over `facedir` is 1) simplicity and 2) you get 6 free bits.
It can be used for things like chests and furnaces and you don't need or want them to "flip them on the side" (like you could with facedir).

`color4dir` is like `colorfacedir`, but you get 64 colors instead of only 8.

Also, as with all color paramtype2s, the builtin functions were updated accordingly.

For details on how this works, see the commits (`lua_api.txt`). Documentation for `facedir` and `wallmounted` is also updated to distinguish those better from `4dir`.

## To do

This PR is ready for review.

Many test cases in DevTest were added which I used for testing in depth. I tested rotation (obviously), color, falling node, and all new nodes I added to DevTest. I also tested the new and modified builtin functions (but you should, too!). It all seems to work to me.

One thing I didn't test in depth is the prediction. I am not so sure if the placement prediction code is correct, please some verify the logic. I don't know how test this part. It would also help me if you have a hint for me on how I can test predition better (I think I basically need to simulate a lag or something ...).

## How to test

Use DevTest and use the many 4dir nodes and the Param2 Tool to modify them. Compare with facedir param2 values 0 to 3, it MUST look the same.
Test if the nodes still look OK when falling (use Falling Node Tool).
Use DevTest and `luacmd` mod to test all the functions I modified.

## Use cases

* Colored and rotatable chests
* Colored and rotatable beds
* Colored and rotatable stairs (as long you don't also want upside-down stairs)

### CODE REVIEW

**Comment by hecktest** [comment]
How about calling it `coloryaw` since it's just around the Y axis?

**Comment by Wuzzy2** [comment]
I don't really care either way. It depends on what others think. If enough people are for it, I change it. If not, then not.

**Comment by Wuzzy2** [comment]
Note: This PR should have a concept approval because #6194 is "supported by core dev".

**Comment by Wuzzy2** [comment]
Bump.

**Comment by SmallJoker** [comment]
Rebase needed.

**Comment by Wuzzy2** [comment]
Rebased.

**Comment by Df458** [comment]
Starting to look at this today. Haven't seen anything too out of the ordinary, but haven't done much with it yet.
From a first look, this seems fine:
4dirs

File: src/client/game.cpp
Comment (Df458) [inline src/client/game.cpp:3472]:
```suggestion
			predicted_f.param_type_2 == CPT2_COLORED_4DIR) {
```

File: src/mapnode.cpp
Comment (Df458) [inline src/mapnode.cpp:244]: Consider a comment or redundant `else if` for 4dir, in case of future paramtype2s

File: src/nodedef.cpp
Comment (Df458) [inline src/nodedef.cpp:1231]:
```suggestion
					features.param_type_2 == CPT2_COLORED_FACEDIR ||
					features.param_type_2 == CPT2_4DIR ||
					features.param_type_2 == CPT2_COLORED_4DIR) {
```

File: src/nodedef.cpp
Comment (Df458) [inline src/nodedef.cpp:1705]: Consider a comment or redundant `else if` for 4dir, in case of future paramtype2s

**Review by Df458 — CHANGES_REQUESTED** [review]

**Comment by Df458** [comment]
Did some further testing, including a test of round-tripping various vectors through `dir_to_fourdir` and `fourdir_to_dir`, no issues on that end.

**Comment by Wuzzy2** [comment]
@Df458: Done.

**Review by Df458 — APPROVED** [review]
Looks good to me now

**Comment by Zughy** [comment]
@Wuzzy2 please rebase

**Comment by Wuzzy2** [comment]
@Zughy: Rebase complete!

**Comment by x2048** [comment]
Tested and reviewed, but a rebase needed. I've pushed rebased branch [here](https://github.com/x2048/minetest/tree/4dir).

@Wuzzy2

**Comment by Wuzzy2** [comment]
Rebase done. Testing needed.

**Comment by Zughy** [comment]
@Wuzzy2 please either remove the label by yourself or quit from being a triager, this is not the first time it happens

**Comment by Wuzzy2** [comment]
Follow-up PR for Minetest Game: #2992

**Pull request merged**
